File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.cc
#include <cstdio>
int N, arr[128][128];
int white = 0;
int blue = 0;
void func(int x, int y, int n){
    bool check = true;
    for(int i = x; i < x + n; i++){
        for(int j = y; j < y + n; j++){
            if(arr[x][y] != arr[i][j]){
                check = false;
                break;
            }
        }
    }
    if(check && arr[x][y] == 0)
        white++;
    else if(check && arr[x][y] == 1)
        blue++;
    else{
        n /= 2;
        func(x, y, n);
        func(x + n, y, n);
        func(x, y + n, n);
        func(x + n, y + n, n);
    }
}
int main(){
    scanf("%d", &N);
    for(int i = 0; i < N; i++){
        for(int j = 0; j < N; j++)
            scanf("%d", &arr[i][j]);
    }
    func(0, 0, N);
    printf("%d\n%d\n", white, blue);
}

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.py
import sys
input = sys.stdin.readline

dx = [1,1,0,0]
dy = [1,0,1,0]

def func(x,y,size):
    global blue, white
    if size == 1:
        if arr[x][y] == 1:
            blue += 1
        else:
            white += 1
        return
    complete = True
    for i in range(size):
        for j in range(size):
            if i == 0 and j == 0: continue
            if arr[x][y] != arr[x+i][y+j]:
                complete = False
                break
    if complete:
        if arr[x][y] == 1:
            blue += 1
        else:
            white += 1
    else:
        dis = size // 2
        for i in range(4):
            func(x + dis*dx[i], y + dis*dy[i], dis)

N = int(input())
arr = []
for _ in range(N):
    arr.append(list(map(int, input().split())))
    
blue = 0
white = 0
func(0,0,N)
print(white)
print(blue)